src/main/java/htsjdk/samtools/SamReaderFactory.java

```diff
@@ -74,11 +76,13 @@
 public abstract class SamReaderFactory {
 
     private static ValidationStringency defaultValidationStringency = ValidationStringency.DEFAULT_STRINGENCY;
-    
+
+    protected Function<SeekableByteChannel, SeekableByteChannel> pathWrapper;
+
     abstract public SamReader open(final File file);
 
     public SamReader open(final Path path) {
-        final SamInputResource r = SamInputResource.of(path);
+        final SamInputResource r = SamInputResource.of(path, getPathWrapper());
         final Path indexMaybe = SamFiles.findIndex(path);
         if (indexMaybe != null) r.index(indexMaybe);
```

Comment:
You're certain the index is always read into memory and cached?  I could have sworn @kcibul was having issues where there was slowdown due to repeatedly accessing the index over the network.  He ended up copying the index files locally.  It looks like there's an option `CACHE_FILE_BASED_INDEXES` maybe he wasn't using that while you are?